src/core/Oscillator.cpp

```diff
@@ -89,6 +108,192 @@ void Oscillator::update( sampleFrame * _ab, const fpp_t _frames,
+	{
+		for (int n = firstBand | 1; n <= bands; n += 2)
+		{
+			table[i] += (1.0f / n) * sinf(F_2PI * i * n / OscillatorConstants::WAVETABLE_LENGTH) / (F_PI / 4);
+		}
+	}
+}
+
+
+
+// Expects waveform converted to frequency domain to be present in the spectrum buffer
+void Oscillator::generateFromFFT(int bands, sample_t* table)
+{
+	// Keep only specified number of bands, set the rest to zero.
+	// Add a +1 offset to the requested number of bands, since the first "useful" frequency falls into bin 1.
+	// I.e., for bands = 1, keeping just bin 0 (center 0 Hz, +- 4 Hz) makes no sense, it would not produce any tone.
+	for (int i = bands + 1; i < OscillatorConstants::WAVETABLE_LENGTH * 2 - bands; i++)
+	{
+		s_specBuf[i][0] = 0.0f;
+		s_specBuf[i][1] = 0.0f;
+	}
+	//ifft
+	fftwf_execute(s_ifftPlan);
+	//normalize and copy to result buffer
+	normalize(s_sampleBuffer, table, OscillatorConstants::WAVETABLE_LENGTH, 2*OscillatorConstants::WAVETABLE_LENGTH + 1);
+}
+
+void Oscillator::generateAntiAliasUserWaveTable(SampleBuffer *sampleBuffer)
+{
+	if (sampleBuffer->m_userAntiAliasWaveTable == NULL) {return;}
+
+	for (int i = 0; i < OscillatorConstants::WAVE_TABLES_PER_WAVEFORM_COUNT; ++i)
+	{
+		for (int i = 0; i < OscillatorConstants::WAVETABLE_LENGTH; ++i)
+		{
+			s_sampleBuffer[i] = sampleBuffer->userWaveSample((float)i / (float)OscillatorConstants::WAVETABLE_LENGTH);
+		}
+		fftwf_execute(s_fftPlan);
+		Oscillator::generateFromFFT(OscillatorConstants::MAX_FREQ / freqFromWaveTableBand(i), (*(sampleBuffer->m_userAntiAliasWaveTable))[i].data());
+	}
+}
+
+
+
+sample_t Oscillator::s_waveTables
+	[Oscillator::WaveShapes::NumWaveShapeTables]
+	[OscillatorConstants::WAVE_TABLES_PER_WAVEFORM_COUNT]
+	[OscillatorConstants::WAVETABLE_LENGTH];
+fftwf_plan Oscillator::s_fftPlan;
+fftwf_plan Oscillator::s_ifftPlan;
+fftwf_complex * Oscillator::s_specBuf;
+float Oscillator::s_sampleBuffer[OscillatorConstants::WAVETABLE_LENGTH];
+
+
+
+void Oscillator::createFFTPlans()
+{
+	Oscillator::s_specBuf = ( fftwf_complex * ) fftwf_malloc( ( OscillatorConstants::WAVETABLE_LENGTH * 2 + 1 ) * sizeof( fftwf_complex ) );
+	Oscillator::s_fftPlan = fftwf_plan_dft_r2c_1d(OscillatorConstants::WAVETABLE_LENGTH, s_sampleBuffer, s_specBuf, FFTW_MEASURE );
+	Oscillator::s_ifftPlan = fftwf_plan_dft_c2r_1d(OscillatorConstants::WAVETABLE_LENGTH, s_specBuf, s_sampleBuffer, FFTW_MEASURE);
+	// initialize s_specBuf content to zero, since the values are used in a condition inside generateFromFFT()
+	for (int i = 0; i < OscillatorConstants::WAVETABLE_LENGTH * 2 + 1; i++)
+	{
+		s_specBuf[i][0] = 0.0f;
+		s_specBuf[i][1] = 0.0f;
+	}
+}
+
+void Oscillator::destroyFFTPlans()
+{
+	fftwf_destroy_plan(s_fftPlan);
+	fftwf_destroy_plan(s_ifftPlan);
+	fftwf_free(s_specBuf);
+}
+
+void Oscillator::generateWaveTables()
+{
+	// Clear all wave tables
+	std::fill(s_waveTables[0][0], s_waveTables[Oscillator::WaveShapes::NumWaveShapeTables][0] - 1, 0.f);
```

Comment:
Two minor issues with that:

1. I get a warning "array index 5 is past the end of the array (which contains 5 elements)"
2. This will not be portable in case anyone will port `s_waveTables` to `std::array` (but it's OK for now)

Reply:
1. Uh, that is annoying. I did that intentionally to avoid a super long line addressing all three dimensions of the array by three long constants. I guess I'll need to do it "by the book" then.. :-/
2. If it is addressed "properly" it should work even for `std::array`, right? It is just a thin wrapper around ordinary array after all..

Reply:
1. You could do something like "roughly" `s_waveTables[Oscillator::WaveShapes::NumWaveShapeTables-1][0] + sizeof(s_waveTables[Oscillator::WaveShapes::NumWaveShapeTables-1] * sizeof(float))` (likely I did that wrong, and it looks a bit ugly now)
2. I guess normally, `std::array` has no further elements, but the C++ designers could add a further member into the array. Though I thought we have an assert somewhere in LMMS that `sizeof(std::array<float,2>)` == `sizeof(float[2])` (#5536), but I can not see it anymore.

Reply:
I changed it to 
```cpp
    std::fill(
        &s_waveTables[0][0][0],
        &s_waveTables[0][0][0] + sizeof s_waveTables / sizeof s_waveTables[0][0][0],
        0.f);
```
and it seems to be working fine, but I don't know, it feels like a hack compared to what I used before (and I'm not sure if your hyperactive compiler won't find a warning in this version as well :smile:). Maybe I should just brute-force it with a triple for loop and be done with it..

Reply:
My compiler says it's OK, but to me it looks hard to say if that's valid.

I guess it would be more elegant using `std::fill(std::begin(s_waveTables[0][0]), std::end(s_waveTables[0][0]), 0.f)`?

Also:

1. We need to do this for all "simple" shapes, not just for the first, right?
2. Don't you start with the highest index and then go down to add more and more harmonics? So `0` is wrong, it should be `WAVE_TABLES_PER_WAVEFORM_COUNT - 1`?

Reply:
Oh, wait, you're trying to initialize the whole table at once... Then I was wrong.

Reply:
Anyways, can't you put

```
std::fill(
    std::begin(s_waveTables[shape][WAVE_TABLES_PER_WAVEFORM_COUNT - 1]),
    std::end(s_waveTables[shape][WAVE_TABLES_PER_WAVEFORM_COUNT - 1]),
    0.f);
```

into `simpleGen`?

Reply:
Ah, I did not even realize that `std::begin` and `std::end` can be used with raw arrays (and mostly I also forgot they exist as independent functions.. I'm still rarely using iterators :) ).

Your last suggestion looks nice and does not waste time on the FFT-based tables, thanks. For some reason, though, the saw wave is now full of aliasing, so I will investigate before pushing this.

Reply:
Hmm, try with index `shape-FirstWaveShapeTable`?

Reply:
Ah, right, I forgot about the enum offset. Btw. found another bug while testing: when the saved project has WT mode enabled, the button looks like it is enabled, but the oscillator is in non-WT mode. Possibly just a missing connection of `modelChanged()` or whatever it is its name, I hope.

Reply:
OK, false alarm. I changed the model default to true in constructor, but not a variable that holds a copy of the model value (since I did not know it existed.

Anyway, the initialization issue should be resolved.